core/store/src/db/colddb.rs

```diff
@@ -0,0 +1,486 @@
+///   of being sorted in the same order the numbers themselves which in turn
+///   means that when we add new heights they are always appended at the end of
+///   the column.
+///
+/// This struct handles all those transformations transparently to the user so
+/// that they don’t need to worry about accessing hot and cold storage
+/// differently.
+///
+/// Note however that this also means that some operations, specifically
+/// iterations, are limited and not implemented for all columns.  When trying to
+/// implement over unsupported column the code will panic.  Refer to particular
+/// iteration method in Database trait implementation for more details.
+///
+/// Lastly, since no data is ever deleted from cold storage, trying to decrease
+/// reference of a value count or delete data is ignored and if debug assertions
+/// are enabled will cause a panic.
+struct ColdDatabase<D: Database = crate::db::RocksDB>(D);
+
+// impl<D: Database + 'static> ColdDatabase<D> {
+//     pub(crate) fn new(db: D) -> std::sync::Arc<dyn Database> {
+//         std::sync::Arc::new(Self(db))
+//     }
+// }
+
+impl<D: Database> ColdDatabase<D> {
+    fn get_impl(&self, col: DBCol, key: &[u8]) -> std::io::Result<Option<Vec<u8>>> {
+        let mut buffer = [0; 32];
+        let key = get_cold_key(col, key, &mut buffer).unwrap_or(key);
+        self.0.get_raw_bytes(col, key)
+    }
+}
+
+impl<D: Database> super::Database for ColdDatabase<D> {
+    fn get_raw_bytes(&self, col: DBCol, key: &[u8]) -> std::io::Result<Option<Vec<u8>>> {
+        let mut result = self.get_impl(col, key);
+        if let Ok(Some(ref mut value)) = result {
+            // Since we’ve stripped the reference count from the data stored in
+            // the database, we need to reintroduce it.  Note that in practice
+            // this should be never called in production since reading of
+            // reference counted columns is done by get_with_rc_stripped.
+            if col.is_rc() {
+                value.extend_from_slice(&1i64.to_le_bytes());
+            }
+        }
+        result
+    }
+
+    fn get_with_rc_stripped(&self, col: DBCol, key: &[u8]) -> std::io::Result<Option<Vec<u8>>> {
+        assert!(col.is_rc());
+        self.get_impl(col, key)
+    }
+
+    /// Iterates over all values in a column.
+    ///
+    /// This is implemented only for a few columns.  Specifically for Block,
+    /// BlockHeader, ChunkHashesByHeight, EpochInfo and StateDlInfos.  It’ll
+    /// panic if used for any other column.
+    ///
+    /// Furthermore, because key of ChunkHashesByHeight is modified in cold
+    /// storage, the order of iteration of that column is different than if it
+    /// would be in hot storage.
+    fn iter<'a>(&'a self, column: DBCol) -> DBIterator<'a> {
+        // Those are the only columns we’re ever iterating over.
+        assert!(
+            matches!(
+                column,
+                DBCol::StateDlInfos
```

Comment:
side comment -- I actually wonder if we shoudl keep StateDlInfo in cold storage at all.

(as AFAIK this is used only when others are asking us for state sync -- which should happen only for the most recent epochs).

Reply:
Yes, you’re right. StadDlInfos doesn’t exist in cold storage.  I just mechanically included all columns that we ever iterate over.